### PR TITLE
Refactor LabelImage to separate out answers from userInput into scoringData

### PR DESCRIPTION
**Pull request by Myranae**
## Summary:
This PR updates LabelImage so that answers are no longer available in the userInput object. This allows the scoring function to have both a userInput parameter and a scoringData parameter to keep answers separate from the user's input with the goal of supporting server side scoring.

PerseusLabelImageMarker and MarkerType contained the same properties. As such, I simplified the code a bit and removed PerseusLabelImageMarker.

In addition, several locations were referencing the wrong types, so those were updated to reference the correct ones. Also, new tests were added confirming the output of `getUserInput` does not contain answers, that `scorePerseusItem` returns the correct results, and that the widget renders correctly if answers are not present in the JSON blob.

Issue: LEMS-2440

## Test plan:
- Confirm all checks pass
- Confirm widget still works as expected

### CODE REVIEW

File: packages/perseus/src/validation.types.ts
Comment (Myranae) [inline packages/perseus/src/validation.types.ts:149]: These are the main changes. Changed `PerseusLabelImageUserInput` so it no longer includes `answers` in the array. Then I moved the parts of `InteractiveMarkerType` needed for scoring to PerseusLabelImageScoringData`.

File: packages/perseus/src/widgets/label-image/label-image.tsx
Comment (Myranae) [inline packages/perseus/src/widgets/label-image/label-image.tsx:76]: Was able to resolve this by removing `PerseusLabelImageMarker`

File: packages/perseus/src/widgets/label-image/score-label-image.ts
Comment (Myranae) [inline packages/perseus/src/widgets/label-image/score-label-image.ts:60]: Data was combined here so that `marker` can have all the available information as `scoreMarker` requires an object of `InteractiveMarkerType`, which includes answers.

File: packages/perseus/src/widgets/label-image/label-image.tsx
Comment (Myranae) [inline packages/perseus/src/widgets/label-image/label-image.tsx:320]: This is a bit of a problem as it requires scoringData on the client-side it seems like. Since this has a TODO associated with it, I left it for now.

Comment (handeyeco) [reply packages/perseus/src/widgets/label-image/label-image.tsx:321]: This is a known issue: https://khanacademy.atlassian.net/browse/LEMS-2544

Luckily AX won't use rationales so we can punt on this for now.

File: packages/perseus/src/validation.types.ts
Comment (handeyeco) [inline packages/perseus/src/validation.types.ts:149]: It's a little higher-risk, but I think we should make this test case pass:

``` TS
    // label-image.test.ts

    describe("getUserInput", () => {
        it("doesn't include answer in getUserInput", async () => {
            // render component
            const {renderer} = renderQuestion(textQuestion);

            const userInput = renderer.getUserInputMap();

            expect(userInput).toEqual({
                "label-image 1": {
                    markers: [
                        {
                            label: "The fourth unlabeled bar line.",
                            x: 25,
                            y: 17.7,
                        },
                        {
                            label: "The third unlabeled bar line.",
                            x: 25,
                            y: 35.3,
                        },
                        {
                            label: "The second unlabeled bar line.",
                            x: 25,
                            y: 53,
                        },
                        {
                            label: "The first unlabeled bar line.",
                            x: 25,
                            y: 70.3,
                        },
                    ],
                },
            });
        });
    });
```

I think there should be another set of tests too:
1. Render a LabelImage widget using ItemData that has no answers (to simulate how it will be on the FE)
2. Answer the LabelImage (both wrong and right)
3. Use `getUserInput` on the Renderer to get the user input
4. Use `scorePerseusItem` to score the user input

More than anything, I'm just surprised that this ticket ended up being more of a types change whereas I was thinking it would end up being more of a logic change. Not saying you're wrong, it just makes me worried we're missing something.

File: .changeset/many-penguins-hug.md
Comment (handeyeco) [inline .changeset/many-penguins-hug.md:2]: Since you changed PerseusTypes in a way that's not backwards compatible, I'd call this a major/breaking change.

File: packages/perseus/src/widgets/label-image/score-label-image.ts
Comment (handeyeco) [reply packages/perseus/src/widgets/label-image/score-label-image.ts:60]: No action needed, but my gut feeling is that this is needless misdirection. I kind of feel we should just change `scoreMarker`:

``` TS
function scoreMarker(
    userInput: APieceOfUserInputType,
    scoringData: APieceOfScoringDataType,
): InteractiveMarkerScore;
```

I'm fine punting on this though.

File: packages/perseus/src/perseus-types.ts
Comment (handeyeco) [inline packages/perseus/src/perseus-types.ts:1062]: @jeremywiebe is particular when it comes to changing `perseus-types.ts`, I would have him sign off first.

**Review by handeyeco — COMMENTED** [review]
I think this is the right direction. I'm wondering though:
1. Will LabelImage render if we provide it with PerseusData sans-`answers`?
2. Will `getUserInput` work sans-`answers`?
Knowing how good you are about manual testing, I'm guessing the answer is "yes." I'd just like to see some automated tests to confirm.

File: packages/perseus/src/widgets/label-image/types.ts
Comment (handeyeco) [inline packages/perseus/src/widgets/label-image/types.ts:2]: These types feel off to me. I'm wondering if it's actually something like:

``` TS
// stuff to render
type LabelImageMarker = {
    // Reveal the correctness state of the user selected answers for the marker.
    showCorrectness: "correct" | "incorrect";
    focused: boolean;
    // Translatable Text; The text to show for the marker. Not displayed directly to the user
    label: string;
    // The marker coordinates on the question image as percent of image size.
    x: number;
    y: number;
}

type LabelImageMarkerScoringData = {
    // The list of correct answers expected for the marker. Often only one but can have multiple
    answers: ReadonlyArray<string>;
}

type LabelImageMarkerUserInput = {
    // The user selected list of answers, used to grade the question.
    selected: ReadonlyArray<string>;
}

type LabelImageFullMarker = LabelImageMarker & LabelImageMarkerScoringData & LabelImageMarkerUserInput;
```

Comment (Myranae) [reply packages/perseus/src/widgets/label-image/types.ts]: Updated! Still need to add recommended tests and do additional manual testing though.

File: packages/perseus/src/widgets/label-image/score-label-image.ts
Comment (Myranae) [reply packages/perseus/src/widgets/label-image/score-label-image.ts]: Fixed!

File: packages/perseus/src/perseus-types.ts
Comment (jeremywiebe) [reply packages/perseus/src/perseus-types.ts:1062]: Further, because TypeScript types match if they are the same shape, I was thinking about ScoringData and ValidationData for each widget as a matching subset of the widget options in this file.

**Review by jeremywiebe — COMMENTED** [review]
Overall I like the direction of this PR. One change I'd like to see is that we keep the type defined in `perseus-types.ts` and then define the ScoringData for it in `validation.types.ts`.

File: packages/perseus/src/perseus-types.ts
Comment (Myranae) [reply packages/perseus/src/perseus-types.ts:1062]: Do you mean use the types in this file to define the ScoringData and ValidationData types?

File: packages/perseus-editor/src/widgets/label-image-editor.tsx
Comment (catandthemachines) [inline packages/perseus-editor/src/widgets/label-image-editor.tsx:34]: Nit: I notice your creating this union in a lot of files. Can you combine them in the types object so you don't have to keep joining them in every file?

**Review by catandthemachines — APPROVED** [review]
The code looks good to me! Though I would encourage you to get an approval from Matthew/Jeremy as they have more context into the scoring code 😉

File: packages/perseus-editor/src/widgets/label-image-editor.tsx
Comment (Myranae) [reply packages/perseus-editor/src/widgets/label-image-editor.tsx:34]: Great suggestion! I'll set that up. :)

File: packages/perseus/src/widgets/label-image/label-image.tsx
Comment (jeremywiebe) [inline packages/perseus/src/widgets/label-image/label-image.tsx:313]: @handeyeco for visibility, Tamara and I re-wrote this function to build a return value by explicitly picking the pieces we need instead of simply return a value out of props. This help stop guarantee we're really only using values in the type and not depending elsewhere on keys that may get stripped when we do the Scoring Data stripping stuff.

File: packages/perseus/src/widgets/label-image/score-label-image.test.ts
Comment (jeremywiebe) [inline packages/perseus/src/widgets/label-image/score-label-image.test.ts:22]: I think we dropped `x` and `y` from this type, didn't we?

File: packages/perseus/src/widgets/label-image/score-label-image.test.ts
Comment (jeremywiebe) [inline packages/perseus/src/widgets/label-image/score-label-image.test.ts:15]: I think we dropped `x` and `y` from this type, didn't we?

File: packages/perseus/src/widgets/label-image/score-label-image.test.ts
Comment (jeremywiebe) [inline packages/perseus/src/widgets/label-image/score-label-image.test.ts:83]: I think once we update this object to only have what's in the user input type, we might not need it. It might be more clear to inline the single `selected: []` bit here instead of spreading this object at that point. I'll leave it to you to decide which way you want to go. :)

**Review by jeremywiebe — APPROVED** [review]

File: packages/perseus/src/widgets/label-image/score-label-image.test.ts
Comment (Myranae) [reply packages/perseus/src/widgets/label-image/score-label-image.test.ts:83]: Updated the tests to remove these empty types.

File: packages/perseus/src/widgets/label-image/__tests__/label-image.test.ts
Comment (Myranae) [reply packages/perseus/src/widgets/label-image/__tests__/label-image.test.ts]: I think the focus _was_ supposed to be making sure answer was not included, but it's also correct to confirm it's just returning the correct userInput, especially since I didn't write any expect statements specifically checking that answers are absent. I do think it's important to emphasize that the goal is to have answers out of user input, but I guess that will be obvious from the shape of UserInput.

File: packages/perseus/src/widgets/label-image/__tests__/label-image.test.ts
Comment (handeyeco) [inline packages/perseus/src/widgets/label-image/__tests__/label-image.test.ts:821]: This is closer but not quite what I was suggesting:

> - Render a LabelImage widget using ItemData that has no answers (to simulate how it will be on the FE)
> - Answer the LabelImage (both wrong and right)
> - Use getUserInput on the Renderer to get the user input
> - Use scorePerseusItem to score the user input

So you're rendering a LabelImage widget without answers, but only asserting that the content around the widget renders (the text you're getting isn't a part of the widget). What I'm proposing is that we repeat "can be answered correctly when correct option is picked for the marker" and "can be answered incorrectly when incorrect option picked for the marker" using `textWithoutAnswersQuestion`.

Comment (Myranae) [reply packages/perseus/src/widgets/label-image/__tests__/label-image.test.ts]: We ended up deciding to handle rendering without answers with future related work. Kept this test to show that rendering doesn't rely on having answer data (the answer arrays are still in the test data, but they are empty. Can modify or remove this test when we start fully handling rendering with no answer property.

**Pull request merged**
